File: mario-rust/src/game/player.rs
use macroquad::prelude::*;

use super::{physics, world::World, Config, InputState};

pub struct Player {
    pub pos: Vec2,
    pub vel: Vec2,
    pub on_ground: bool,
    size: Vec2,
    facing: f32,
    coyote_timer: f32,
    jump_buffer_timer: f32,
    powered: bool,
    invuln_timer: f32,
}

impl Player {
    pub fn new(spawn: Vec2, config: &Config) -> Self {
        let size = config.player_size;
        let pos = spawn + vec2((config.tile_size - size.x) * 0.5, config.tile_size - size.y);

        Self {
            pos,
            vel: Vec2::ZERO,
            on_ground: false,
            size,
            facing: 1.0,
            coyote_timer: 0.0,
            jump_buffer_timer: 0.0,
            powered: false,
            invuln_timer: 0.0,
        }
    }

    pub fn reset(&mut self, spawn: Vec2, config: &Config) {
        let size = config.player_size;
        self.pos = spawn + vec2((config.tile_size - size.x) * 0.5, config.tile_size - size.y);
        self.vel = Vec2::ZERO;
        self.on_ground = false;
        self.facing = 1.0;
        self.coyote_timer = 0.0;
        self.jump_buffer_timer = 0.0;
        self.powered = false;
        self.invuln_timer = 0.0;
        self.size = size;
    }

    pub fn update(&mut self, input: &InputState, world: &World, config: &Config, dt: f32) -> bool {
        self.invuln_timer = (self.invuln_timer - dt).max(0.0);
        let mut jumped = false;
        if input.jump_pressed {
            self.jump_buffer_timer = config.jump_buffer_time;
        } else {
            self.jump_buffer_timer = (self.jump_buffer_timer - dt).max(0.0);
        }

        if input.jump_released && self.vel.y < 0.0 {
            self.vel.y *= config.jump_cut_multiplier;
        }

        if self.on_ground {
            self.coyote_timer = config.coyote_time;
        } else {
            self.coyote_timer = (self.coyote_timer - dt).max(0.0);
        }

        if input.move_x.abs() > f32::EPSILON {
            self.facing = input.move_x.signum();
        }

        let target_speed = input.move_x * config.move_speed;
        let accel = if input.move_x.abs() > f32::EPSILON {
            config.move_accel
        } else {
            config.move_decel
        };
        self.vel.x = physics::approach(self.vel.x, target_speed, accel * dt);

        if self.jump_buffer_timer > 0.0 && self.coyote_timer > 0.0 {
            self.vel.y = -config.jump_speed;
            self.on_ground = false;
            self.coyote_timer = 0.0;
            self.jump_buffer_timer = 0.0;
            jumped = true;
        }

        self.vel.y = (self.vel.y + config.gravity * dt).min(config.terminal_velocity);

        let (pos, vel, on_ground) =
            physics::move_with_collisions(self.pos, self.size, self.vel, &world.solids, dt);

        self.pos = pos;
        self.vel = vel;
        self.on_ground = on_ground;

        if self.jump_buffer_timer > 0.0 && self.on_ground {
            self.vel.y = -config.jump_speed;
            self.on_ground = false;
            self.coyote_timer = 0.0;
            self.jump_buffer_timer = 0.0;
            jumped = true;
        }

        jumped
    }

    pub fn size(&self) -> Vec2 {
        self.size
    }

    pub fn center(&self) -> Vec2 {
        self.pos + self.size * 0.5
    }

    pub fn rect(&self) -> Rect {
        physics::rect_at(self.pos, self.size)
    }

    pub fn facing_dir(&self) -> f32 {
        self.facing
    }

    pub fn is_powered(&self) -> bool {
        self.powered
    }

    pub fn set_powered(&mut self, powered: bool) {
        self.powered = powered;
    }

    pub fn is_invulnerable(&self) -> bool {
        self.invuln_timer > 0.0
    }

    pub fn start_invulnerability(&mut self, duration: f32) {
        self.invuln_timer = duration.max(0.0);
    }
}

File: mario-cpp/core/src/player.cpp
#include "mario/core/player.hpp"

#include <algorithm>

#include "mario/core/constants.hpp"
#include "mario/core/physics.hpp"
#include "mario/core/types.hpp"

namespace mario::core {

void Player::reset(Vec2 spawn_tile, const Config& config) {
  size = config.player_size;
  pos = Vec2{
      spawn_tile.x + (config.tile_size - size.x) / 2,
      spawn_tile.y + (config.tile_size - size.y),
  };

  vel = Vec2{0, 0};
  on_ground = false;
  facing = 1;
  coyote_timer = 0;
  jump_buffer_timer = 0;
  powered = false;
  invuln_timer = 0;
}

bool Player::update(const StepInput& input, const World& world, const Config& config) {
  invuln_timer = std::max<std::int32_t>(0, invuln_timer - kDtTimeUnits);

  bool jumped = false;

  if (input.jump_pressed) {
    jump_buffer_timer = config.jump_buffer_time;
  } else {
    jump_buffer_timer = std::max<std::int32_t>(0, jump_buffer_timer - kDtTimeUnits);
  }

  if (input.jump_released && vel.y < 0) {
    vel.y /= 2;  // jump_cut_multiplier = 0.5
  }

  if (on_ground) {
    coyote_timer = config.coyote_time;
  } else {
    coyote_timer = std::max<std::int32_t>(0, coyote_timer - kDtTimeUnits);
  }

  const int move_x = input.move_x();
  if (move_x != 0) {
    facing = move_x < 0 ? -1 : 1;
  }

  const units_t target_speed = static_cast<units_t>(move_x) * config.move_speed;
  const units_t accel = (move_x != 0) ? config.move_accel : config.move_decel;
  vel.x = physics::approach(vel.x, target_speed, accel);

  if (jump_buffer_timer > 0 && coyote_timer > 0) {
    vel.y = -config.jump_speed;
    on_ground = false;
    coyote_timer = 0;
    jump_buffer_timer = 0;
    jumped = true;
  }

  vel.y = std::min(vel.y + config.gravity, config.terminal_velocity);

  const physics::MoveResult moved = physics::move_with_collisions(pos, size, vel, world.solids);
  pos = moved.pos;
  vel = moved.vel;
  on_ground = moved.on_ground;

  if (jump_buffer_timer > 0 && on_ground) {
    vel.y = -config.jump_speed;
    on_ground = false;
    coyote_timer = 0;
    jump_buffer_timer = 0;
    jumped = true;
  }

  return jumped;
}

}  // namespace mario::core

